core/src/main/scala/org/elasticmq/FifoDeduplicationIdsHistory.scala

```diff
@@ -0,0 +1,74 @@
+package org.elasticmq
+
+import com.typesafe.scalalogging.LazyLogging
+import org.elasticmq.FifoDeduplicationIdsHistory.deduplicationInterval
+import org.elasticmq.actor.queue.InternalMessage
+import org.elasticmq.util.NowProvider
+import org.joda.time.DateTime
+
+import scala.annotation.tailrec
+
+/** Contains history of used Deduplication IDs associated with incoming messages to FIFO queues
+  * @param messagesByDeduplicationId contains all registered deduplication IDs with associated messages. Used as a fast access storage for lookups if given ID was already registered
+  * @param deduplicationIdsByCreationDate Deduplication IDs stored together with the message creation date.
+  *                                       Incoming IDs should be already sorted by their creation date so it is safe to assume that the list will be ordered from oldest to newest.
+  *                                       Used for fast lookups for messages by their creation date while cleaning outdated messages
+  */
+case class FifoDeduplicationIdsHistory(
+    messagesByDeduplicationId: Map[DeduplicationId, InternalMessage],
+    deduplicationIdsByCreationDate: List[DeduplicationIdWithCreationDate]
+) extends LazyLogging {
+
+  def addNew(message: InternalMessage): FifoDeduplicationIdsHistory = {
+    message.messageDeduplicationId match {
+      case Some(deduplicationId) if !messagesByDeduplicationId.contains(deduplicationId) =>
+        new FifoDeduplicationIdsHistory(
+          messagesByDeduplicationId + ((deduplicationId, message)),
+          deduplicationIdsByCreationDate :+ DeduplicationIdWithCreationDate(deduplicationId, message.created)
+        )
+      case _ => this
+    }
+  }
+
+  def wasRegistered(maybeDeduplicationId: Option[DeduplicationId]): Option[InternalMessage] =
+    maybeDeduplicationId.flatMap(deduplicationId => messagesByDeduplicationId.get(deduplicationId))
+
+  def cleanOutdatedMessages(nowProvider: NowProvider): FifoDeduplicationIdsHistory = {
+    val (idsToRemove, notTerminatedMessages) =
+      partitionMapUntil(deduplicationIdsByCreationDate)(
+        map = _.id,
+        cond = _.creationDate.plusMinutes(deduplicationInterval).isBefore(nowProvider.now)
```

Comment:
[minor] we could capture the `now` once, to avoid getting the system time on each loop

Reply:
Yeah, I even thought about it and put it into "backlog" of things to do, but in the end forgot about it :P Changed to invoke `now` only once